File: typescript/packages/var-core/src/cell-diff.ts
import type { Row, Table } from './ast.ts'
import type { Span } from './span.ts'

// One checked column of one header-bound row: the input the comparison needs.
export type RowCheck = {
  readonly column: string
  readonly value: string // the cell text, e.g. "9"
  readonly span: Span // the cell text's source range in the .md
}

// The verdict for one checked column after comparing against the table.
export type CellDiff = {
  readonly column: string
  readonly span: Span
  readonly expected: string
  readonly actual: string
  readonly ok: boolean
  // The raw pre-display values, present on the inline-parameter path (where
  // comparison is deep equality over transformed values). Adapters hand them
  // to their test framework's structural differ; they are never serialized
  // into run results or conformance artifacts.
  readonly expectedValue?: unknown
  readonly actualValue?: unknown
  // True when the parameter type's `format` rendered `actual` — the display
  // pair is document notation, and adapters should prefer it over the raw
  // values in their expected/actual projection.
  readonly formatted?: boolean
}

// Display rules 2–4 of the mismatch-rendering chain (rule 1, the parameter
// type's `format`, applies only on the inline-parameter path — see
// param-diff.ts): a string as-is, any other primitive stringified, anything
// else as best-effort JSON. The JSON fallback is port-native and deliberately
// outside conformance — bundles that pin an object actual must use `format`.
export function renderCellValue(value: unknown): string {
  if (typeof value === 'string') return value
  if (typeof value !== 'object' || value === null) return String(value)
  try {
    return JSON.stringify(value) ?? String(value)
  } catch {
    return String(value)
  }
}

// Compare a row step's returned object against the row's cells. Only columns
// present on `returned` are checked; the rest are inputs. A non-object return
// (including undefined) checks nothing.
export function compareRow(
  returned: unknown,
  checks: ReadonlyArray<RowCheck>,
): ReadonlyArray<CellDiff> {
  if (returned === null || typeof returned !== 'object') return []
  const obj = returned as Record<string, unknown>
  const diffs: CellDiff[] = []
  for (const check of checks) {
    if (!(check.column in obj)) continue
    const actual = renderCellValue(obj[check.column])
    diffs.push({
      column: check.column,
      span: check.span,
      expected: check.value,
      actual,
      ok: actual === check.value,
    })
  }
  return diffs
}

// Thrown by the executor when a header-bound row's returned columns don't all
// match. Carries the mismatched cells so adapters render/record them.
export class CellMismatchError extends Error {
  readonly cells: ReadonlyArray<CellDiff>
  constructor(cells: ReadonlyArray<CellDiff>) {
    super(cells.map((c) => `${c.column}: expected ${c.expected} but was ${c.actual}`).join('; '))
    this.name = 'CellMismatchError'
    this.cells = cells
  }
}

export function isCellMismatchError(e: unknown): e is CellMismatchError {
  return e instanceof CellMismatchError
}

// The step returned the wrong TYPE (a non-array where a table was input, a
// string where a doc string was input) or wrong SHAPE (row/column count,
// missing record key, mixed row forms). An author mistake, not a value diff.
export class ReturnShapeError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ReturnShapeError'
  }
}

// Compare a whole-table step's returned table against the input table, full
// reproduction: every column of every data row is checked (the header row is
// labels, never compared). `returned` may be an array-of-arrays (data rows,
// positional) or an array-of-records (keyed by header cell). Cells compare as
// exact strings (`String(value) === cellText`). `undefined` → no checks.
// Type/shape problems throw `ReturnShapeError`.
export function compareTable(returned: unknown, input: Table): ReadonlyArray<CellDiff> {
  if (returned === undefined) return []
  if (!Array.isArray(returned)) {
    throw new ReturnShapeError(`expected a table (array of rows), got ${typeof returned}`)
  }
  const columns = input.header.cells
  const dataRows = input.rows
  if (returned.length !== dataRows.length) {
    throw new ReturnShapeError(`expected ${dataRows.length} row(s), got ${returned.length}`)
  }
  const isRecord = (r: unknown): r is Record<string, unknown> =>
    r !== null && typeof r === 'object' && !Array.isArray(r)
  const allArrays = returned.every((r) => Array.isArray(r))
  const allRecords = returned.every(isRecord)
  if (!allArrays && !allRecords) {
    throw new ReturnShapeError('table rows must be all arrays or all objects')
  }
  const diffs: CellDiff[] = []
  for (let i = 0; i < dataRows.length; i++) {
    const row = dataRows[i] as Row
    const ret = returned[i]
    if (allArrays) {
      const cells = ret as ReadonlyArray<unknown>
      if (cells.length !== columns.length) {
        throw new ReturnShapeError(
          `row ${i}: expected ${columns.length} column(s), got ${cells.length}`,
        )
      }
    }
    for (let j = 0; j < columns.length; j++) {
      const column = columns[j] as string
      let actualValue: unknown
      if (allArrays) {
        actualValue = (ret as ReadonlyArray<unknown>)[j]
      } else {
        const rec = ret as Record<string, unknown>
        if (!(column in rec)) {
          throw new ReturnShapeError(`row ${i}: missing column "${column}"`)
        }
        actualValue = rec[column]
      }
      const expected = row.cells[j] ?? ''
      const actual = renderCellValue(actualValue)
      diffs.push({
        column,
        span: row.cellSpans[j] ?? row.span,
        expected,
        actual,
        ok: actual === expected,
      })
    }
  }
  return diffs
}

File: java/var-core/src/main/java/dev/varar/core/CellDiff.java
package dev.varar.core;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Table row/cell comparison — port of {@code var-core/src/cell-diff.ts}.
 *
 * <p>{@code CellDiff} itself is the verdict for one checked column after comparing a step's
 * returned value against the authored Markdown cells: {@link #compareRow} produces these for a
 * header-bound row (only the columns the step actually returns are checked), {@link
 * #compareTable} for a whole reproduced table (every column of every data row is checked).
 *
 * <p>A step's returned value has no static shape in this hexagonal core (it crosses from
 * user-authored step handlers), so {@code returned} is {@code Object} here and is narrowed via
 * {@code instanceof Map}/{@code instanceof List} — the same duck-typing {@code cell-diff.ts} does
 * with {@code typeof}/{@code Array.isArray}. A returned "row object" is therefore expected as a
 * {@link Map} (keyed by column name); a returned "table" as a {@link List} of {@link List}
 * (positional) or a {@link List} of {@link Map} (keyed by header cell) — mirroring the {@code
 * Map<String, Object>} convention already used for JSON-object-shaped values elsewhere in this
 * port (see {@code CanonicalJson}).
 */
public record CellDiff(
        String column,
        Span span,
        String expected,
        String actual,
        boolean ok,
        Object expectedValue,
        Object actualValue,
        boolean formatted) {

    /**
     * The pre-{@code format} raw values ({@code expectedValue}/{@code actualValue}) are
     * populated on the inline-parameter path (where comparison is deep equality over
     * transformed values — see {@link ParamDiff#compareParams}) so adapters can hand them
     * to their test framework's structural differ; {@code formatted} is true when the
     * parameter type's {@code format} rendered {@code actual}, telling adapters to prefer
     * the document-notation display pair over the raw values. Neither is ever serialized
     * into run results or conformance artifacts ({@code Conformance.failureCell} projects
     * only {@code column}/{@code expected}/{@code actual}/{@code span}). This convenience
     * constructor keeps the row/table comparison paths (and their tests) on the original
     * five components, with {@code null} raw values.
     */
    public CellDiff(String column, Span span, String expected, String actual, boolean ok) {
        this(column, span, expected, actual, ok, null, null, false);
    }

    /** One checked column of one header-bound row: the input the comparison needs. */
    public record RowCheck(String column, String value, Span span) {}

    /**
     * Display rules 2–4 of the mismatch-rendering chain (rule 1, the parameter type's
     * {@code format}, applies only on the inline-parameter path — see {@link ParamDiff}):
     * a string as-is, any other primitive/boxed primitive via {@link String#valueOf}, and
     * anything else as a best-effort {@link String#valueOf} — the port-native fallback
     * that is deliberately outside conformance (bundles that pin an object-valued actual
     * must give the parameter type a {@code format}).
     */
    public static String renderCellValue(Object value) {
        if (value instanceof String s) return s;
        return String.valueOf(value);
    }

    /**
     * Compares a row step's returned value against the row's cells. Only columns present on
     * {@code returned} (expected to be a {@link Map}) are checked; the rest are inputs. A
     * non-map return (including {@code null}) checks nothing.
     */
    public static List<CellDiff> compareRow(Object returned, List<RowCheck> checks) {
        if (!(returned instanceof Map<?, ?> obj)) return List.of();
        List<CellDiff> diffs = new ArrayList<>();
        for (RowCheck check : checks) {
            if (!obj.containsKey(check.column())) continue;
            String actual = renderCellValue(obj.get(check.column()));
            diffs.add(new CellDiff(check.column(), check.span(), check.value(), actual, actual.equals(check.value())));
        }
        return List.copyOf(diffs);
    }

    /**
     * Thrown by the executor when a header-bound row's returned columns don't all match. Carries
     * the mismatched cells so adapters render/record them.
     */
    public static final class CellMismatchException extends RuntimeException {
        private final List<CellDiff> cells;

        public CellMismatchException(List<CellDiff> cells) {
            super(message(cells));
            this.cells = List.copyOf(cells);
        }

        public List<CellDiff> cells() {
            return cells;
        }

        private static String message(List<CellDiff> cells) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < cells.size(); i++) {
                if (i > 0) sb.append("; ");
                CellDiff c = cells.get(i);
                sb.append(c.column())
                        .append(": expected ")
                        .append(c.expected())
                        .append(" but was ")
                        .append(c.actual());
            }
            return sb.toString();
        }
    }

    /** Type-guard helper mirroring {@code isCellMismatchError} in {@code cell-diff.ts}. */
    public static boolean isCellMismatchException(Throwable e) {
        return e instanceof CellMismatchException;
    }

    /**
     * The step returned the wrong TYPE (a non-list where a table was input, a non-string where a
     * doc string was input) or wrong SHAPE (row/column count, missing record key, mixed row
     * forms). An author mistake, not a value diff.
     */
    public static final class ReturnShapeException extends RuntimeException {
        public ReturnShapeException(String message) {
            super(message);
        }
    }

    /**
     * Compares a whole-table step's returned table against the input table, full reproduction:
     * every column of every data row is checked (the header row is labels, never compared).
     * {@code returned} may be a {@code List} of {@code List} (data rows, positional) or a {@code
     * List} of {@code Map} (keyed by header cell). Cells compare as exact strings ({@code
     * renderCellValue(value).equals(cellText)}). {@code null} → no checks. Type/shape problems
     * throw {@link ReturnShapeException}.
     */
    public static List<CellDiff> compareTable(Object returned, Ast.Table input) {
        if (returned == null) return List.of();
        if (!(returned instanceof List<?> rows)) {
            throw new ReturnShapeException("expected a table (array of rows), got " + typeName(returned));
        }
        List<String> columns = input.header().cells();
        List<Ast.Row> dataRows = input.rows();
        if (rows.size() != dataRows.size()) {
            throw new ReturnShapeException("expected " + dataRows.size() + " row(s), got " + rows.size());
        }
        boolean allArrays = rows.stream().allMatch(r -> r instanceof List<?>);
        boolean allRecords = rows.stream().allMatch(r -> r instanceof Map<?, ?>);
        if (!allArrays && !allRecords) {
            throw new ReturnShapeException("table rows must be all arrays or all objects");
        }
        List<CellDiff> diffs = new ArrayList<>();
        for (int i = 0; i < dataRows.size(); i++) {
            Ast.Row row = dataRows.get(i);
            Object ret = rows.get(i);
            if (allArrays) {
                List<?> cells = (List<?>) ret;
                if (cells.size() != columns.size()) {
                    throw new ReturnShapeException(
                            "row " + i + ": expected " + columns.size() + " column(s), got " + cells.size());
                }
            }
            for (int j = 0; j < columns.size(); j++) {
                String column = columns.get(j);
                Object actualValue;
                if (allArrays) {
                    actualValue = ((List<?>) ret).get(j);
                } else {
                    Map<?, ?> rec = (Map<?, ?>) ret;
                    if (!rec.containsKey(column)) {
                        throw new ReturnShapeException("row " + i + ": missing column \"" + column + "\"");
                    }
                    actualValue = rec.get(column);
                }
                String expected = j < row.cells().size() ? row.cells().get(j) : "";
                String actual = renderCellValue(actualValue);
                Span span = j < row.cellSpans().size() ? row.cellSpans().get(j) : row.span();
                diffs.add(new CellDiff(column, span, expected, actual, actual.equals(expected)));
            }
        }
        return List.copyOf(diffs);
    }

    private static String typeName(Object o) {
        return o.getClass().getSimpleName();
    }
}
